sw/banshee/src/tran.rs

```diff
@@ -2135,9 +2135,20 @@ impl<'a> InstructionTranslator<'a> {
             riscv::OpcodeRdRs1Rs2::Sll => LLVMBuildShl(self.builder, rs1, rs2, name),
             riscv::OpcodeRdRs1Rs2::Srl => LLVMBuildLShr(self.builder, rs1, rs2, name),
             riscv::OpcodeRdRs1Rs2::Sra => LLVMBuildAShr(self.builder, rs1, rs2, name),
-            riscv::OpcodeRdRs1Rs2::Dmcpy => self
-                .section
-                .emit_call("banshee_dma_strt", [self.dma_ptr(), rs1, rs2]),
+            riscv::OpcodeRdRs1Rs2::Dmcpy => self.section.emit_call(
+                "banshee_dma_strt",
+                [
+                    self.dma_ptr(),
+                    LLVMBuildBitCast(
+                        self.builder,
+                        self.section.state_ptr,
+                        LLVMPointerType(LLVMInt8Type(), 0),
+                        NONAME,
+                    ),
+                    rs1,
+                    rs2,
+                ],
+            ),
```

Comment:
Do you know why the CPU pointer is an 8-bit pointer here? I think we could clean this up with the following changes just to make it a bit easier to understand.

```diff
diff --git a/sw/banshee/src/runtime/jit.ll b/sw/banshee/src/runtime/jit.ll
index 07bedce..d1f8903 100644
--- a/sw/banshee/src/runtime/jit.ll
+++ b/sw/banshee/src/runtime/jit.ll
@@ -34,7 +34,7 @@ declare void @banshee_ssr_eoi(%SsrState* %ssr)
 
 declare void @banshee_dma_src(%DmaState* writeonly %dma, i32 %lo, i32 %hi)
 declare void @banshee_dma_dst(%DmaState* writeonly %dma, i32 %lo, i32 %hi)
-declare i32 @banshee_dma_strt(%DmaState* %dma, i8* %cpu, i32 %size, i32 %flags)
+declare i32 @banshee_dma_strt(%DmaState* %dma, %Cpu* %cpu, i32 %size, i32 %flags)
 declare i32 @banshee_dma_stat(%DmaState* readonly %dma, i32 %addr)
 
 declare i32* @banshee_reg_ptr(%Cpu* %cpu, i32 %reg)
diff --git a/sw/banshee/src/tran.rs b/sw/banshee/src/tran.rs
index c825f41..a244e33 100644
--- a/sw/banshee/src/tran.rs
+++ b/sw/banshee/src/tran.rs
@@ -1348,17 +1348,7 @@ impl<'a> InstructionTranslator<'a> {
         let value = match data.op {
             riscv::OpcodeImm5RdRs1::Dmcpyi => self.section.emit_call(
                 "banshee_dma_strt",
-                [
-                    self.dma_ptr(),
-                    LLVMBuildBitCast(
-                        self.builder,
-                        self.section.state_ptr,
-                        LLVMPointerType(LLVMInt8Type(), 0),
-                        NONAME,
-                    ),
-                    rs1,
-                    imm,
-                ],
+                [self.dma_ptr(), self.section.state_ptr, rs1, imm],
             ),
             // _ => bail!("Unsupported opcode {}", data.op),
         };
@@ -2137,17 +2127,7 @@ impl<'a> InstructionTranslator<'a> {
             riscv::OpcodeRdRs1Rs2::Sra => LLVMBuildAShr(self.builder, rs1, rs2, name),
             riscv::OpcodeRdRs1Rs2::Dmcpy => self.section.emit_call(
                 "banshee_dma_strt",
-                [
-                    self.dma_ptr(),
-                    LLVMBuildBitCast(
-                        self.builder,
-                        self.section.state_ptr,
-                        LLVMPointerType(LLVMInt8Type(), 0),
-                        NONAME,
-                    ),
-                    rs1,
-                    rs2,
-                ],
+                [self.dma_ptr(), self.section.state_ptr, rs1, rs2],
             ),
             _ => bail!("Unsupported opcode {}", data.op),
         };
```

Reply:
Nice catch, thanks! I just copy-pasted it and didn't notice that this could be simplified.